cirq-core/cirq/ops/common_gate_families.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Common Gate Families used in cirq-core"""
+
+from typing import cast, Optional, Type, Union
+
+from cirq.ops import gateset, raw_types, parallel_gate, eigen_gate
+from cirq import protocols
+
+
+class AnyUnitaryGateFamily(gateset.GateFamily):
+    def __init__(self, num_qubits: Optional[int] = None) -> None:
+        self._num_qubits = num_qubits
+        if num_qubits:
+            name = f'{num_qubits}-Qubit UnitaryGateFamily'
+            description = f'Accepts any {num_qubits}-qubit unitary gate.'
+        else:
+            name = 'Any-Qubit UnitaryGateFamily'
+            description = 'Accepts any unitary gate.'
+        super().__init__(raw_types.Gate, name=name, description=description)
+
+    def _predicate(self, g: raw_types.Gate) -> bool:
+        return (
+            self._num_qubits is None or protocols.num_qubits(g) == self._num_qubits
+        ) and protocols.has_unitary(g)
+
+    def __repr__(self) -> str:
+        return f'cirq.AnyUnitaryGateFamily(num_qubits = {self._num_qubits})'
+
+
+class AnyIntegerPowerGateFamily(gateset.GateFamily):
+    def __init__(self, gate: Type[eigen_gate.EigenGate]) -> None:
+        if not (isinstance(gate, type) and issubclass(gate, eigen_gate.EigenGate)):
+            raise ValueError(f'{gate} must be a subclass of `cirq.EigenGate`.')
+        super().__init__(
+            gate,
+            name=f'AnyIntegerPowerGateFamily: {gate}',
+            description=f'Accepts any instance `g` of `{gate}` s.t. `g.exponent` is an integer.',
+        )
```

Comment:
this might somewhat unituitively reject:

```python3
cirq.google.SycamoreGate() ** 2 # becomes an fsim later
```
Could we name this `IntegerPowerEigenGateFamily` or something ?

Reply:
We won't be able to initialize this with `Sycamore` / `Fsim` as the constructor checks that input gate should be a subclass of `EigenGate`. 
I can rename it to `AnyIntegerPowerEigenGateFamily` but I'd rather prefer adding documentation which conveys that this can be only be used with Eigen gates (and ofc, the verification check already exists)

Reply:
added documentation.